### PR TITLE
[REEF-1002] Add Vortex examples that cause large serialization cost

### PR DESCRIPTION
**Pull request by yunseong**
This addressed the issue by
  * Adding matrix multiplication example based on @johnyangk's code.
  * Defining user-defined Input/output to cause serialization cost.

JIRA:
  [REEF-1002](https://issues.apache.org/jira/browse/REEF-1002)

### CODE REVIEW

**Comment by yunseong** [comment]
This PR will not conflict with #675 because it does not change `VortexFunction`. @afchung Could you take a look? By the way, the callback you implemented works fine and is very useful. Thanks. :)

File: lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/matmul/MatMulStart.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * MatMul User Code Example.
+ * This example multiplies two matrices by distributing computation to multiple Tasklets.
+ * Each Tasklet receives split of the matrix on the left side, and copy of the matrix on the right side.
+ * To check whether the results are correct, Identity matrix is used for the matrix on the right side.
+ */
+final class MatMulStart implements VortexStart {
```
Comment (afchung) [inline lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/matmul/MatMulStart.java]: Nit: rename to ``IdentityMatMulStart`` to make it more explicit.

File: lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/matmul/RowMatrix.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+ */
+final class RowMatrix implements Matrix<Double> {
+  private final List<Vector<Double>> values;
+
+  /**
+   * Constructor of matrix which creates an empty matrix of size (numRow x numColumn).
+   * @param values Elements of Matrix.
+   */
+  RowMatrix(final List<Vector<Double>> values) {
+    this.values = values;
```
Comment (afchung) [inline lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/matmul/RowMatrix.java]: You might want to deep-copy here to make immutable.

File: lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/matmul/RowMatrix.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Vector;
+
+/**
+ * Row-oriented matrix implementation used in {@link MatMul} example.
+ */
+final class RowMatrix implements Matrix<Double> {
+  private final List<Vector<Double>> values;
```
Comment (afchung) [inline lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/examples/matmul/RowMatrix.java]: Just curious, but why not ``List<List<>>``? Do we need the synchronization property that ``Vector<>`` provides?

**Comment by afchung** [comment]
@yunseong This looks good to me in general, but I would suggest usage of [``unmodifiableList``](http://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#unmodifiableList(java.util.List)) to ensure immutability of matrices.

**Comment by yunseong** [comment]
@afchung Thanks for the review and comments! I've addressed your comments. While doing this, I've fixed some bugs in `RowMatrix` in `add()` and `multiply()`. Also, to reduce copying cost, each sub-matrix on the left side is generated at once, which was done in two steps: `generate-then-split`.
Could you take another look? Thanks a lot!

**Comment by afchung** [comment]
Thanks @yunseong, I'll take a second look.

**Comment by afchung** [comment]
Looks good to me. Should anyone else take a look?

**Comment by bgchun** [comment]
@afchung Thanks.
@yunseong The Travis CI build failed. Did you take a look? 

Adding @johnyangk as well. What happens when there is an exception while executing `VortexFunction`? How does Vortex handle this case? I'm guessing `VortexFuture` should have a way to tell it.

**Comment by johnyangk** [comment]
@bgchun 

We have `TOutput call(TInput input) throws Exception` in `VortexFunction`. This means that user code can throw an exception inside the call. The exception is caught in the Evaluator and forwarded to Driver, same as how the result is forwarded.

You are right. The exception is surfaced to the user via `TOutput get() throws InterruptedException, ExecutionException` in `VortexFuture` as an `ExecutionException`.

**Comment by bgchun** [comment]
Thanks. Does callback has an API to receive an exception? 
If the client registers a callback function, it's not gonna call `get`.

**Comment by yunseong** [comment]
@bgchun Travis tests pass in JDK 7, but failed in JDK 8. The reason was `WatcherTest`, which was reported to the [dev mailing list](http://mail-archives.apache.org/mod_mbox/incubator-reef-dev/201511.mbox/%3CCAFu6yH3a1FfRCxtHkCTz_p7r2xop1sYJ_yvav747MDTRDvx-pA@mail.gmail.com%3E). @gwsshs22 is looking at it.

**Comment by yunseong** [comment]
@bgchun I didn't consider the exceptional cases: I thought the callback is optional for successful case. Client still can catch exception in `VortexFuture.get()`, but I think it would be great to send failure event via callback as well. @afchung @johnyangk  Any other thought?

**Comment by johnyangk** [comment]
We can maybe register an exception handler as well via that particular `VortexThreadPool#submit`.

This is just my opinion, but how about we do not return a `VortexFuture` from it so that there's is only one way of getting tasklet result/exception for each `VortexThreadPool#submit`?

**Comment by zpqian** [comment]
FYI the API design in .NET threadpool:
https://msdn.microsoft.com/en-us/library/hh194741.aspx

**Comment by afchung** [comment]
Perhaps, instead of ``EventHandler``, we can use ``[Observer](https://github.com/apache/incubator-reef/blob/master/lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/rx/Observer.java)`` for the callback instead, and invoke ``onError`` on ``Exceptions``. What does everyone else think?

**Comment by bgchun** [comment]
Yes. It's better to use a variation of Observer. Since Observer has a
specific contract (zero or more on Next calls followed by onCompleted or
onError), we can define a subclass of Observer like ExactlyOnceObserver
(better name?), which calls either onNext or onError. onCompleted should
raise an UnsupportedOperationException. How does this sound?


On Wed, Nov 25, 2015 at 1:38 PM, Andrew Chung <notifications@github.com>
wrote:

> Perhaps, instead of EventHandler, we can use [Observer](
> https://github.com/apache/incubator-reef/blob/master/lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/rx/Observer.java
> ) for the callback instead, and invoke onError on Exceptions. What does
> everyone else think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/676#issuecomment-159490376>
> .
>



-- 
Byung-Gon Chun

**Comment by bgchun** [comment]
@gwsshs22 Did you have a chance to look at the Travis CI problem?

**Comment by bgchun** [comment]
@yunseong If the client uses `callback`, it's not going to call `get`.

**Comment by yunseong** [comment]
@bgchun I totally agree with you. Exceptions should be passed to the client even with callbacks. 
Also, `ExactlyOnceObserver` sounds great to me. Maybe we need a discussion via mailing list or JIRA?

**Comment by bgchun** [comment]
Can you open a JIRA issue? Since it's specific to Vortex Future, we don't
need to discuss via our mailing list.

On Sun, Nov 29, 2015 at 5:10 PM, Yunseong Lee <notifications@github.com>
wrote:

> @bgchun <https://github.com/bgchun> I totally agree with you. Exceptions
> should be passed to the client even with callbacks.
> Also, ExactlyOnceObserver sounds great to me. Maybe we need a discussion
> via mailing list or JIRA?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/676#issuecomment-160390870>
> .
>



-- 
Byung-Gon Chun

**Comment by yunseong** [comment]
I opened REEF-1023(https://issues.apache.org/jira/browse/REEF-1023).

What about @johnyangk's suggestion that having only one way to get the result/exception of `VortexFuture`? I think sometimes user might want to use `VortexFuture.get()` for simplicity. Any other opinion?

**Comment by bgchun** [comment]
`callback` when the client wants non-blocking submits. `get` when the client wants blocking submits. 
If the client uses `get`, the client usually does not pass a callback function. This's typical.

@yunseong @johnyangk What is your concern?

**Comment by bgchun** [comment]
@yunseong We can merge this PR now, and then you can change this example to address the callback issue. Or we can wait for the callback issue to be addressed, update this PR, and merge it afterwards. How would you like to proceed?

**Comment by yunseong** [comment]
@bgchun I just wanted to bring up what @johnyangk suggested. I think it's natural to have both methods.

For merging this, either way is fine with me. But this change has not been merged yet, so I think it'd be better to merge this after the callback issue is addressed.

**Comment by bgchun** [comment]
Sounds good. Thanks.

On Sun, Nov 29, 2015 at 9:35 PM, Yunseong Lee <notifications@github.com>
wrote:

> @bgchun <https://github.com/bgchun> I just wanted to bring up what
> @johnyangk <https://github.com/johnyangk> suggested. I think it's natural
> to have both methods.
>
> For merging this, either way is fine with me. But this change has not been
> merged yet, so I think it'd be better to merge this after the callback
> issue is addressed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/676#issuecomment-160414040>
> .
>



-- 
Byung-Gon Chun

**Comment by afchung** [comment]
@yunseong if we are waiting for REEF-1023, perhaps we should close this PR first such that no one accidentally merges it.

**Comment by yunseong** [comment]
@afchung Thanks for the suggestion. i'll close this for now, and revisit later.

**Pull request closed without merging**